### PR TITLE
fix: documentation + refactoring of methods + add default classes

### PR DESCRIPTION
**Pull request by abdullah-cognite**
Refactoring everything in dotnet-simulator-utils and exposing Default Classes and methods that connectors should use, instead of directly using classes from utils to build connectors.

### CODE REVIEW

File: Cognite.Simulator.Utils/ConnectorBase.cs [hunk 1]
```diff
@@ -124,27 +125,26 @@ public SimulatorIntegration GetSimulatorIntegration(string simulator)
...
-        /// <see cref="EnsureSimulatorIntegrationsExist(CancellationToken)"/>
+        /// <see cref="InitRemoteSimulatorIntegrations(CancellationToken)"/>
         /// </summary>
         /// <param name="token">Cancellation token</param>
         public abstract Task Init(CancellationToken token);
         
         /// <summary>
         /// Implements the connector loop. Should call the <see cref="Heartbeat(CancellationToken)"/> method and any
         /// other thats that are done periodically by the connector
+        /// TODO: We should be able to control the Run method implemented by the connector so that we can handle exceptions thrown internally
```
Comment (sagar-thalwar) [inline Cognite.Simulator.Utils/ConnectorBase.cs:146]: Looks like this is not a TODO

File: Cognite.Simulator.Utils/ConnectorBase.cs [hunk 2]
```diff
@@ -196,16 +185,16 @@ public virtual TimeSpan GetHeartbeatInterval()
         /// How often to report the simulator license status information back to CDF (Heartbeat)
         /// </summary>
         /// <returns>Time interval</returns>
-        public virtual TimeSpan GetLicenseCheckInterval()
+        public TimeSpan GetLicenseCheckInterval()
         {
-            int min3600 = _config.LicenseCheck.Interval < 3600 ? 3600 : _config.LicenseCheck.Interval;
-            return TimeSpan.FromSeconds(min3600);
+            int interval = _config.LicenseCheck.Interval < ONE_HOUR ? ONE_HOUR : _config.LicenseCheck.Interval;
```
Comment (sagar-thalwar) [inline Cognite.Simulator.Utils/ConnectorBase.cs:190]: Good refactor to not hard code numbers and move them to constants

File: Cognite.Simulator.Utils/ConnectorBase.cs [hunk 3]
```diff
@@ -338,15 +327,15 @@ protected async Task EnsureSimulatorIntegrationsExist(CancellationToken token)
...
+        public async Task HeartbeatLoop(CancellationToken token)
         {
             while (!token.IsCancellationRequested)
             {
                 await Task
                     .Delay(GetHeartbeatInterval(), token)
                     .ConfigureAwait(false);
                 _logger.LogDebug("Updating connector heartbeat");
-                await UpdateSimulatorIntegrations(false, token)
+                await this.UpdateRemoteSimulatorIntegrations(false, token)
```
Comment (sagar-thalwar) [inline Cognite.Simulator.Utils/ConnectorBase.cs:338]: Why `this` ?

File: Cognite.Simulator.Utils/ConnectorBase.cs [hunk 4]
```diff
@@ -375,16 +364,16 @@ await Task
                     .ConfigureAwait(false);
                 _logger.LogDebug("Updating connector license timestamp");
                 LastLicenseCheckTimestamp = DateTime.UtcNow.ToUnixTimeMilliseconds();
-                LastLicenseCheckResult = CheckLicenseStatus() ? LicenseStatus.AVAILABLE : LicenseStatus.NOT_AVAILABLE;
-                await UpdateSimulatorIntegrations(false, token)
+                LastLicenseCheckResult = GetLicenseStatus() ;
+                await this.UpdateRemoteSimulatorIntegrations(false, token)
```
Comment (sagar-thalwar) [inline Cognite.Simulator.Utils/ConnectorBase.cs:368]: why `this` ?

File: Cognite.Simulator.Utils/DefaultClasses/DefaultConnector.cs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+            }
+            catch (OperationCanceledException)
+            {
+                _logger.LogWarning("Cancellation requested. Exiting the connector");
+            }
+            catch (AggregateException aggregateException)
+            {
+                foreach (var innerEx in aggregateException.InnerExceptions)
+                {
+                    if (innerEx is NewConfigDetected)
```
Comment (sagar-thalwar) [inline Cognite.Simulator.Utils/DefaultClasses/DefaultConnector.cs:117]: Nice work, maybe some comments?

File: Cognite.Simulator.Utils/DefaultClasses/DefaultConnectorRuntime.cs [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        }
+
+    }
+
+    public static async Task Run(ILogger defaultLogger, CancellationToken token)
+    {
+        var assembly = Assembly.GetEntryAssembly();
+        var services = new ServiceCollection();
+        services.AddLogger();
+        services.AddCogniteClient("CalculatorConnector", "CalculatorConnector/v0.0.1 (Cognite)", true);
```
Comment (sagar-thalwar) [inline Cognite.Simulator.Utils/DefaultClasses/DefaultConnectorRuntime.cs:59]: Can we update this to use a dynamic string coming from a client?

File: Cognite.Simulator.Utils/DefaultClasses/DefaultConnectorRuntime.cs [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+        DefaultConfig<TAutomationConfig> config;
+        try
+        {
+            config = await services.AddConfiguration<DefaultConfig<TAutomationConfig>>(
+                path: "./config.yml",
+                types: new Type[] { typeof(DefaultConnectorConfig), typeof(SimulatorConfig) },
+                appId: "CalculatorConnector",
```
Comment (sagar-thalwar) [inline Cognite.Simulator.Utils/DefaultClasses/DefaultConnectorRuntime.cs:67]: appId should be dynamic

File: Cognite.Simulator.Utils/DefaultClasses/DefaultConnectorRuntime.cs [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+        catch (ConfigurationException e)
+        {
+            defaultLogger.LogError("Failed to load configuration file: {Message}", e.Message);
+            return;
+        }
+
+        services.AddStateStore();
+        services.AddHttpClient<FileStorageClient>();
+        services.AddScoped<TAutomationConfig>();
+        ConfigureServices?.Invoke(services);
```
Comment (sagar-thalwar) [inline Cognite.Simulator.Utils/DefaultClasses/DefaultConnectorRuntime.cs:80]: This allows clients to be injected into services. Maybe comments?

File: Cognite.Simulator.Utils/DefaultClasses/DefaultConnectorRuntime.cs [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+                    {
+                        // If the simulator is not available, log to Windows events and exit
+                        defaultLogger.LogError(sue.Message);
+                        logger.LogError(sue, sue.Message);
+                        break;
+                    }
+                    else
+                    {
+                        logger.LogError(e, "Unhandled exception: {message}", e.Message);
+                        // sourceProgram.Cancel();
```
Comment (sagar-thalwar) [inline Cognite.Simulator.Utils/DefaultClasses/DefaultConnectorRuntime.cs:179]: Can we find out what this commented code does?

File: Cognite.Simulator.Utils/ModelLibraryBase.cs [hunk 10]
```diff
@@ -647,7 +644,7 @@ public interface IModelProvider<T>
         /// </summary>
         /// <param name="modelExternalId">Model external id</param>
         /// <returns>List of state objects</returns>
-        IEnumerable<T> GetAllModelRevisions(string modelExternalId);
+        //IEnumerable<T> GetAllModelRevisions(string modelExternalId);
```
Comment (sagar-thalwar) [inline Cognite.Simulator.Utils/ModelLibraryBase.cs:647]: We can remove commented code

File: Cognite.Simulator.Utils/ModelParsingInfo.cs [hunk 11]
```diff
@@ -39,6 +39,7 @@ public class ModelParsingInfo
 
     /// <summary>
     /// Extension utilities for the model parsing info
+    /// TODO: move the methods to model parsing info
```
Comment (sagar-thalwar) [inline Cognite.Simulator.Utils/ModelParsingInfo.cs:42]: Possible to address the TODO in this PR?

File: Cognite.Simulator.Utils/ModelStateBase.cs [hunk 12]
```diff
@@ -8,6 +8,7 @@ namespace Cognite.Simulator.Utils
 {
     /// <summary>
     /// This base class represents the state of a model file
+    /// TODO: See if we can remove FileState completely and move all the variables into this class
```
Comment (sagar-thalwar) [inline Cognite.Simulator.Utils/ModelStateBase.cs:11]: Can we create a JIRA ticket for this TODO and add the reference to ticket here?

**Review by sagar-thalwar — COMMENTED** [review]
Some minor comments. Overall great work!

File: Cognite.Simulator.Utils/DefaultClasses/DefaultConnectorRuntime.cs (hunk 9 above)
Comment (abdullah-cognite) [reply Cognite.Simulator.Utils/DefaultClasses/DefaultConnectorRuntime.cs:179]: Seems to be part of old dead code (the variable does not exist), so deleted it

**Review by sagar-thalwar — APPROVED** [review]
🚀

**Pull request merged**